Unwatch users when updates are no longer needed

mathiascode opened the pull request:
Currently, the list of watched users grows indefinitely until we disconnect/quit. Unwatch users when we no longer need to track them.

slook commented on pynicotine/interests.py:
```diff
@@ -134,6 +144,15 @@ def request_similar_users(self):
+        self.similar_users.clear()
+
         for username in msg.users:
+            rating = msg.users[username] if isinstance(msg.users, dict) else None
```
`msg.users` should always be a `dict` even if it is empty, so `isinstance()` shouldn't be required.

mathiascode replied:
It can be a list or dict depending on which server message it is (one doesn't include ratings). I'll probably come up with something better.

slook commented on pynicotine/users.py:
```diff
@@ -28,14 +28,16 @@
         self.upload_speed = None
         self.files = None
         self.folders = None
+        self.modes = set()
```
String matching to identify different modes isn't very nice.

Consider defining a new class such as `WatchMode()` for this purpose.

mathiascode replied:
It doesn't feel right to make the users module responsible for knowing every mode in the application. Ideally you could call watch_user() from anywhere (say, a plugin) without being required to add a new mode here.

slook replied:
Perhaps get a value without needing to pass an argument with something in https://docs.python.org/3/library/inspect.html#module-inspect

slook commented on pynicotine/users.py:
```diff
@@ -28,14 +28,16 @@
         self.files = None
         self.folders = None
+        self.modes = set()
+        self.implicit = True  # True if no WatchUser message was sent (i.e. receiving updates for room user)
```
Consider a `set()` to contain all the `rooms` that a user is in, avoids doing string concatenations.

If the set is empty the user would not be in any of the rooms we are in.

mathiascode replied:
Similar concerns as the previous comment. I don't think the users module should know this.

mathiascode replied:
I'll get rid of `mode.startswith("chatroom_")` in favor of a boolean parameter for watch_user().

slook commented on pynicotine/buddies.py:
```diff
@@ -165,7 +165,7 @@ def add_buddy(self, username):
 
         # Request user status, speed and number of shared files
-        core.users.watch_user(username)
+        core.users.watch_user(username, mode="userlist")
```
Ambiguity with normal user. "userlist" -> "buddy"

slook replied:
Rather than "buddy", actually "buddies" would match what is used in `core.enabled_components`.

mathiascode commented:
I renamed mode -> context, seemed a little clearer to me.

Also refactored transfer user watching so it doesn't unwatch/rewatch when changing the state of the last remaining transfer.

slook commented on pynicotine/users.py:
```diff
@@ -116,20 +118,86 @@ def request_set_status(self, status):
+            self.watched[username] = watched_user = WatchedUser(username)
+
+        if not context:
+            context = "unknown"
```
Why is this needed?

If the function call is made improperly, then an error message should be logged.

mathiascode replied:
I'd like to get the change into 3.3.x, but wasn't sure if any plugins are calling this function. I'll just mark watch_user() with no context as deprecated, and make the context mandatory in 3.4.0.

slook commented on pynicotine/users.py:
```diff
@@ -116,20 +118,86 @@ def request_set_status(self, status):
+            return
+
+        if not context:
+            context = "unknown"
```
If the context is unspecified, then perhaps all contexts ought to be discarded?

mathiascode replied:
Too easy to misuse and break things. I'll make the context mandatory, since it's a new function.
